Mesh coarsening generator

GiudGiud opened the pull request:
Kind of a side project but I think it's valuable.
There's only one piece missing to get it to work on HEX afaict. 
EDIT
hex done

GiudGiud committed:
Factor out detection-of-refined-noncoformal-neighbors diagnostic code into a util for coarsening

GiudGiud committed:
Add a mesh generator that can coarsen the mesh
refs #25280

GiudGiud committed:
Add documentation for mesh coarsening
refs #25280

GiudGiud committed:
Add error testing and the basic for single and multiple coarsenings tests
refs #25280

GiudGiud committed:
- add verbose logs
- fix issues with elements in candidate list being deleted already
refs #25280

GiudGiud committed:
Test three configurations for multiple block coarsenings:
- uniform
- non uniform 0 and 1 level of refinement
- non uniform 0 to 2 levels
refs #25280

GiudGiud committed:
Switch to a starting point for the coarsening as it can be used at multiple levels of refinement
refs #25280

GiudGiud committed:
Remove some debugging logs
Remove accomodations for libmesh h-refined elements: we cannot coarsen as we cannot delete half of a refined element
Less unnecessary preparing for use
refs #25280

GiudGiud committed:
Keep subdomain ids while coarsening
refs #25280

GiudGiud committed:
Add a heuristic to be able to coarsen inside a subdomain but not all the subdomain
Start from the starting point, stop if elements are roughly half the size of a coarse element
refs #25280

GiudGiud committed:
Add tests for coarsening meshes that were non-uniformly refined at boundaries
refs #25280

GiudGiud committed:
Non unity build fixes
Test adjustment
Fixup a doco link
Ban parallel execution until it is fixed

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1992586/) on 7d7961d wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/26445/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1997636/) on 8a2fc94 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>2758a7</th>
      <th colspan=4>#26445 8a2fc9</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.08%</td>
      <td>85.09%</td>
      <td>+0.01%</td>
      <td>92.83%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>98600</td>
      <td>98795</td>
      <td>+195</td>
      <td>220</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>17294</td>
      <td>17309</td>
      <td>+15</td>
      <td>17</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/26445/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/26445/coverage/framework)


## Modules coverage

Coverage did not change

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/26445/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/26445/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/26445/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/26445/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/26445/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/26445/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/26445/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/26445/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/26445/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/26445/coverage/geochemistry)
- [`heat_transfer`](https://mooseframework.inl.gov/docs/PRs/26445/coverage/heat_transfer)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/26445/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/26445/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/26445/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/26445/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/26445/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/26445/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/26445/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/26445/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/26445/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/26445/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/26445/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/26445/coverage/scalar_transport)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/26445/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/26445/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/26445/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/26445/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/26445/coverage/xfem)

</details>



This comment will be updated on new commits.

GiudGiud commented on framework/src/meshgenerators/CoarsenBlockGenerator.C:
```diff
@@ -0,0 +1,411 @@
+    // we should attempt to form coarse element from
+    // TODO: think about the implications of set vs vector. Set might grow to the entire mesh
+    // due to sorting. Vector we could insert at the beginning and treat new candidates immediately
+    std::set<std::pair<Elem *, Node *>> candidate_pairs;
```
because this will sort pointers, it wont take the same path in parallel I think, hence why this cannot run in parallel even replicated

If I added a custom comparator for ordering of the set it would fix that. I also dont think we need to run this in parallel

loganharbour replied:
You can add a custom comparator

lindsayad replied:
You can look at `AutomaticMortarGeneration` which leverages `CompareDofObjectsByID` in order to get consistent sorting in parallel. You would have to implement it for the pair, but it's the same idea

GiudGiud replied:
Added sorting along a direction.

GiudGiud replied:
Direction will be nicer than ID ordering as in the future I can see users specify it to help the algorithm coarsen. It works fine with a random one for now

GiudGiud replied:
Worked with even more sets of pointers for 3D and running into these ordering issues again.

I dont really set the need for replicated parallel mesh generation. Especially for such an expensive step where it make sense to serialize or distribute but not really replicate. 

I would be down to move forward with serial here.

lindsayad replied:
It's convenient to be able to run a single input for mesh generation and problem solve. I think that's probably the main argument for parallel replicated mesh generation

GiudGiud replied:
yeah it is. I gave it another shot.

Turns out custom comparators are tricky. You have to return a boolean, but it wont complain if you return a Real.
Should work now

GiudGiud committed:
Add an ordering of candidates points for future coarsening

lindsayad commented on framework/src/meshgenerators/CoarsenBlockGenerator.C:
```diff
@@ -0,0 +1,420 @@
+  // Take ownership of the mesh
+  std::unique_ptr<MeshBase> mesh = std::move(_input);
+  if (!mesh->is_replicated())
+    paramError("input", "Input mesh must be distributed");
```
```suggestion
    paramError("input", "Input mesh must not be distributed");
```

lindsayad commented on framework/doc/content/source/meshgenerators/CoarsenBlockGenerator.md:
```diff
@@ -0,0 +1,41 @@
+
+## Overview
+
+This MeshGenerator object allows the user to coarsen one or more blocks in a mesh. It has been designed
+to work with previously uniformly refined meshes, though the heuristic can still work in the general case.
```
This documentation combined with what is happening in the source code really confused me. Based on this documentation I would have expected that this generator only works if the mesh has been refined, e.g. that element levels are > 0. But instead it is the opposite
```c++
      if (current_elem->level() > 0)
        mooseError("H-refined meshes cannot be coarsened with this mesh generator. Use the "
                   "[Adaptivity] block to coarsen them.");
```
It would be a nice symmetry if `CoarsenBlockGenerator` reused the code from `RefineBlockGenerator` but instead set `Elem::COARSEN` instead of `Elem::REFINE`.

Of course that will be a different functionality than what you are actually doing in your code here. You'll need to change these sentences to not confuse people. I might even suggest a different name for the class although I could be persuaded otherwise from a user perspective. From a libMesh developer perspective, I have a hard time not expecting `CoarsenBlockGenerator` to be the coarsening analog, and potentially reusing the code, of `RefineBlockGenerator`

GiudGiud replied:
> It would be a nice symmetry if CoarsenBlockGenerator reused the code from RefineBlockGenerator but instead set Elem::COARSEN instead of Elem::REFINE.

sure but it also would not be very useful since we could only work with meshes refined with libmesh AMR (and that 'need' is fully covered with initial adaptivity steps)

I ll try to modify the sentence

GiudGiud committed:
Add more disclaimers to docs for coarsening capabilities

GiudGiud committed:
Extend mesh coarsening utils to handle hex elements

GiudGiud committed:
Fix logic for breaking out of loops
Add more 3D support in coarsening generator

GiudGiud committed:
Add tests for coarsening of 3D hexes

GiudGiud commented:
Added 3D support, same tests essentially.

lindsayad commented on framework/doc/content/source/meshgenerators/CoarsenBlockGenerator.md:
```diff
@@ -0,0 +1,49 @@
+
+The user has to provide the ids of the blocks to be modified, as well as the corresponding levels of coarsening for each block. These must match up to the order of the aforementioned block id list, e.g. if the blocks you wish to modify are '0 2 1 4', with 1 coarsening for block 0, 2 for block 2, 3 for block 1, and 4 for block 4, then the coarsening list will need to look like '1 2 3 4'.
+
+The user provides the location of starting point for the coarsening algorithm with the [!param](/Mesh/CoarsenBlockGenerator/starting_point) parameter. The element containing that point is found, then the algorithm attempts to coarsen by selecting each node of that element as the interior node of a coarse element containing that element. Once this element is coarsened, the neighbors of the coarse  ...
```
```suggestion
The user provides the starting point for the coarsening algorithm with the [!param](/Mesh/CoarsenBlockGenerator/starting_point) parameter. The element containing that point is found, then the algorithm attempts to coarsen by selecting each node of that element as the interior node of a coarse element containing that element. Once this element is coarsened, the neighbors of the coarse element
```

lindsayad commented on framework/doc/content/source/meshgenerators/CoarsenBlockGenerator.md:
```diff
@@ -0,0 +1,49 @@
+The user has to provide the ids of the blocks to be modified, as well as the corresponding levels of coarsening for each block. These must match up to the order of the aforementioned block id list, e.g. if the blocks you wish to modify are '0 2 1 4', with 1 coarsening for block 0, 2 for block 2, 3 for block 1, and 4 for block 4, then the coarsening list will need to look like '1 2 3 4'.
+
+The user provides the location of starting point for the coarsening algorithm with the [!param](/Mesh/CoarsenBlockGenerator/starting_point) parameter. The element containing that point is found, then the algorithm attempts to coarsen by selecting each node of that element as the interior node of a coarse element containing that element. Once this element is coarsened, the neighbors of the coarse  ...
+are considered as targets for the next round of coarsening, until all elements are coarsened once for that round of coarsening. If
```
How does this work with the block restriction? E.g. what happens if the starting point is in a subdomain that is surrounded by blocks not specified for coarsening, but there are other blocks specified for coarsening? I looked and saw that it is not a `std::vector<Point>` for the `starting_point` parameter, but just a single `Point`

GiudGiud replied:
if there is no contiguous path to reaching those blocks from the starting_point, you are out of luck.

A vector of starting points would be fine to me

GiudGiud replied:
It's an evolution compared to what we have now. I had in mind trying to detect automatically where to start instead, in future works

GiudGiud replied:
Actually this wont work.,

If you pick a single starting point, there are only 4 choices for the central nodes of the coarse elements.
If you pick multiple starting points, it's 4x the choices for every other starting point.

lindsayad commented on framework/doc/content/source/meshgenerators/CoarsenBlockGenerator.md:
```diff
@@ -0,0 +1,49 @@
+!alert note
+If the input mesh has non-conformalities due to prior use of adaptive mesh refinement,
+this mesh generator may be able to remove them.
+However if the input mesh is disjoint, this will not be improved. You may try a [MeshRepairGenerator.md]
```
Can you define what `disjoint` means?

GiudGiud replied:
added more details

GiudGiud committed:
Address Alex's review
- improve doco

Co-authored-by: Alex Lindsay <alexander.lindsay@inl.gov>

GiudGiud committed:
Make coarsening algorithm more predictable in 3D:
- sorting sets deterministically

Exit early if fine element types do not match as we dont know how to get the opposite node from the interior node

lindsayad commented on framework/doc/content/source/meshgenerators/CoarsenBlockGenerator.md:
```diff
@@ -0,0 +1,54 @@
+
+## Overview
+
+This MeshGenerator object allows the user to coarsen one or more blocks in a mesh. It has been designed
+to work with previously refined meshes that were saved to a file without the mesh refinement pattern information. For example, refined meshes stored using [Exodus.md] or [VTK.md] file formats do not contain the information needed to recognize parent and children elements from uniform refinement. The heuristic can still work in the general case of non-uniform refinement.
```
So the first thing I tried was changing your 2x2 block in `coarsen_quad.i` to 3x3 and I get the result I expected to get which is hanging nodes. I think that more often than not this generator will not work with meshes that were not generated through uniform block refinement. I would consider dropping this last sentence; if you don't want to drop it, then an example of a non-uniform case in which this will succeed might be good.

I think that ideally the generator should error if hanging nodes are generated

GiudGiud replied:
the internal sideset examples are non-uniform refinement (not uniform over the whole domain, or even a whole block)

I can put the mesh diagnostics as a sub-generator if you d like

GiudGiud replied:
> more often than not 

Absolutely in the infinite space of all meshes.
In the also infinite but smaller space of AMR-refined meshes for which we lost the adaptivity information, this is where it needs to shine

lindsayad replied:
Can you describe the use case for this? Previously you said

> sure but it also would not be very useful since we could only work with meshes refined with libmesh AMR

in reference to a coarsening generator that worked in the same way as the current refinement generator but now you are saying that the intent of your object is to work with AMR-refined meshes. But for your object we need to add an additional input file and insert an additional step of writing out the mesh to exodus or vtk format. Are there a lot of people doing AMR with third-party meshing software before loading the mesh into MOOSE?

GiudGiud replied:
> I can put the mesh diagnostics as a sub-generator if you d like. This sub-generator can be set to error on the presence of non-conformalities. It will need to be switchable on/off as sometimes we will need two rounds of coarsening to get two separate refined regions coarsened

actually i dont think I can do mesh diagnostics as a sub-generator.
But I could move the code that detects non-conformality to a MeshDiagnosticsUtil.C and call it from the new generator? What do you think?

GiudGiud commented:
so an update here, I've deployed this on a sample mesh I wanted to fix and found out I cant get the mesh to output entirely without polyhedrals. Filtering out the polyhedrals in paraview gets me a mesh with holes that is useless. There's not a lot of hope there until we can read polyhedrals

lindsayad commented:
so with the update, is there anything this object does that couldn't be done by saving the mesh to checkpoint format and then doing the coarsening analog of `RefineBlockGenerator`?

GiudGiud commented:
Even when we have polyhedral support we still need to either undo or recognize/translate to libmesh the external AMR.

lindsayad commented:
I'm basically just asking if you think this PR is still evaluable. If you do, then I will give a thorough review next week

GiudGiud commented on framework/src/utils/MeshCoarseningUtils.C:
```diff
@@ -0,0 +1,273 @@
+getFineElementFromInteriorNode(const libMesh::Node * const interior_node,
+                               const libMesh::Node * const reference_node,
+                               const libMesh::Elem * const elem,
+                               const libMesh::Real /*non_conformality_tol*/,
```
```suggestion
```

GiudGiud commented on framework/src/utils/MeshCoarseningUtils.C:
```diff
@@ -0,0 +1,273 @@
+bool
+getFineElementFromInteriorNode(const libMesh::Node * const interior_node,
+                               const libMesh::Node * const reference_node,
+                               const libMesh::Elem * const elem,
```
should probsbly switch those to references

GiudGiud replied:
done

GiudGiud commented on framework/src/utils/MeshCoarseningUtils.C:
```diff
@@ -0,0 +1,273 @@
+
+    // Pick a node (mid-face for the coarse element) to form the base
+    // We must use the same element reproducibly, despite the pointers being ordered in the set
+    const Elem * one_fine_elem = nullptr;
```
```suggestion
    // We use the element id to choose the same element consistently
    const Elem * one_fine_elem = nullptr;
```

GiudGiud commented on framework/src/utils/MeshCoarseningUtils.C:
```diff
@@ -0,0 +1,273 @@
+      }
+    const auto interior_node_index = one_fine_elem->get_node_index(interior_node);
+
+    // Find a side which contains the interior node
```
```suggestion
    // Find any side which contains the interior node
```

GiudGiud commented on framework/src/utils/MeshCoarseningUtils.C:
```diff
@@ -0,0 +1,273 @@
+        an_interior_node_side = s;
+        break;
+      }
+    // The node we seek is on the same side, but opposite from the interior node
```
```suggestion
    // A node near a face of the coarse element seek is on the same side, but opposite from the interior node
```

GiudGiud commented on framework/src/utils/MeshCoarseningUtils.C:
```diff
@@ -0,0 +1,273 @@
+        one_fine_elem->opposite_node(interior_node_index, an_interior_node_side);
+    const auto center_face_node = one_fine_elem->node_ptr(center_face_node_index);
+
+    // The exterior nodes are the opposite nodes of the interior_node!
```
```suggestion
    // We gather the coarse element nodes from the fine elements that share the center face node we just selected
```

GiudGiud commented on framework/src/utils/MeshCoarseningUtils.C:
```diff
@@ -0,0 +1,273 @@
+        other_fine_elems.push_back(neighbor);
+        continue;
+      }
+      const auto interior_node_number = neighbor->get_node_index(interior_node);
```
```suggestion
      // The coarse element node is the opposite nodes of the interior_node in a fine element
      const auto interior_node_number = neighbor->get_node_index(interior_node);
```

GiudGiud commented on framework/src/utils/MeshCoarseningUtils.C:
```diff
@@ -0,0 +1,273 @@
+    }
+
+    // Center face node was not shared with 4 elements
+    // We could try again on any of the 5 other coarse element faces
```
```suggestion
    // We could try again on any of the 5 other coarse element faces but we don't insist for now
```

GiudGiud commented on framework/src/utils/MeshCoarseningUtils.C:
```diff
@@ -0,0 +1,273 @@
+        }
+      mooseAssert(fine_elem, "Search for fine element should have worked");
+
+      // Find the other fine element opposite that element (shares a side)
```
```suggestion
      // Find the other fine element opposite the element containing the node (they share a side)
```

GiudGiud commented on framework/include/utils/MeshCoarseningUtils.h:
```diff
@@ -0,0 +1,52 @@
+bool getFineElementFromInteriorNode(const libMesh::Node * const interior_node,
+                                    const libMesh::Node * const reference_node,
+                                    const libMesh::Elem * const elem,
+                                    const libMesh::Real non_conformality_tol,
```
```suggestion
```

GiudGiud commented on framework/src/meshgenerators/CoarsenBlockGenerator.C:
```diff
@@ -0,0 +1,445 @@
+                                                      "The list of blocks to be coarsened");
+  params.addRequiredParam<std::vector<unsigned int>>(
+      "coarsening",
+      "Minimum amount of times to coarsen each block, corresponding to their index in 'block'");
```
```suggestion
      "Maximum amount of times to coarsen elements in each block. See 'block' for indexing");
```

GiudGiud commented on framework/src/meshgenerators/CoarsenBlockGenerator.C:
```diff
@@ -0,0 +1,445 @@
+        if (!fine_elem)
+          continue;
+
+        // If we are using libmesh h-refinement before coarsening (for testing), we cannot just
+        // leave the parent in the mesh so we delete it too
+        if (fine_elem->parent() && !parent_deleted)
+        {
+          parent_deleted = true;
+          mesh_copy->delete_elem(fine_elem->parent());
+        }
```
```suggestion
```

GiudGiud commented:
> I'm basically just asking if you think this PR is still evaluable. If you do, then I will give a thorough review next week

I would say don't prioritize it at all. But yes there's value

GiudGiud committed:
Address self-review

- better comments, docstrings
- use reference over pointers for utility arguments

Update framework/src/meshgenerators/CoarsenBlockGenerator.C

Update framework/src/meshgenerators/CoarsenBlockGenerator.C

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2030603/) on 38d75e2 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/26445/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/26445/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 796cb8aa68a15d61eb82eb353bb6582b01c6634f`

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2030807/) on 126da44 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/26445/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/26445/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 796cb8aa68a15d61eb82eb353bb6582b01c6634f`

lindsayad commented on framework/include/utils/MeshCoarseningUtils.h:
```diff
@@ -0,0 +1,51 @@
+/**
+ * Utility routine to gather vertex nodes for, and elements contained in, for a coarse QUAD or HEX
+ * element
+ * @param interior_node the node inside the coarse element
+ * @param elem an element containing the node. Its neighbor lists must be up to date so it
```
doxygen for `reference_node`

lindsayad commented on framework/include/utils/MeshCoarseningUtils.h:
```diff
@@ -0,0 +1,51 @@
+ * @param tentative_coarse_nodes nodes to be used to form the coarse element
+ * @param fine_elements fine elements that are inside the coarse element
+ */
+bool getFineElementFromInteriorNode(const libMesh::Node & interior_node,
```
```suggestion
bool getFineElementsFromInteriorNode(const libMesh::Node & interior_node,
```

lindsayad commented on framework/include/utils/MeshCoarseningUtils.h:
```diff
@@ -0,0 +1,51 @@
+ * Utility routine to get the index of the node opposite, in the element, to the node
+ * of interest
+ * @param elem_type type of the element
+ * @param node_index element index of the node of interest
```
```suggestion
 * @param node_index local index of the node of interest
```

lindsayad commented on framework/include/utils/MeshCoarseningUtils.h:
```diff
@@ -0,0 +1,51 @@
+ * of interest
+ * @param elem_type type of the element
+ * @param node_index element index of the node of interest
+ * @return element index of the opposite node, with regards to the element
```
```suggestion
 * @return local index of the opposite node, with regards to the element
```

lindsayad commented on framework/src/meshgenerators/CoarsenBlockGenerator.C:
```diff
@@ -0,0 +1,432 @@
+
+  // Take ownership of the mesh
+  std::unique_ptr<MeshBase> mesh = std::move(_input);
+  if (!mesh->is_replicated())
```
```suggestion
  if (!mesh->is_serial())
```

lindsayad commented on framework/src/meshgenerators/CoarsenBlockGenerator.C:
```diff
@@ -0,0 +1,432 @@
+CoarsenBlockGenerator::recursiveCoarsen(const std::vector<subdomain_id_type> & block_ids,
+                                        std::unique_ptr<MeshBase> & mesh,
+                                        const std::vector<unsigned int> & coarsening,
+                                        unsigned int max,
```
```suggestion
                                        const unsigned int max,
```

lindsayad commented on framework/src/meshgenerators/CoarsenBlockGenerator.C:
```diff
@@ -0,0 +1,432 @@
+      const auto sorting =
+          (a.first->vertex_average() - b.first->vertex_average()) * sorting_direction;
+      if (MooseUtils::absoluteFuzzyGreaterThan(sorting, 0))
+        return 1;
+      else if (MooseUtils::absoluteFuzzyEqual(sorting, 0) &&
+               MooseUtils::absoluteFuzzyGreaterThan((*a.second - *b.second) * sorting_direction, 0))
+        return 1;
+      else
+        return -1;
```
All of these implicitly convert to `true`. You can try
```c++
#include <iostream>

int
main()
{
  const bool value = -1;
  std::cout << "The value is " << value << std::endl;
}
```
The output is 1

GiudGiud replied:
best comparator ever

GiudGiud committed:
Apply suggestions from code review
- fix candidate ordering comparator
- fix docstrings

Co-authored-by: Alex Lindsay <alexander.lindsay@inl.gov>

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2035280/) on 459cc7f wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/26445/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/26445/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 796cb8aa68a15d61eb82eb353bb6582b01c6634f`

GiudGiud commented:
@lindsayad please let me know the path forward here.
If you want the generator to error if the coarsening is not good for the mesh, I ll refactor the diagnostics into a utils outside of the MeshDiagnosticsGenerator. This will incidentally get us closer to having a `[Diagnostics]` system that can re-use those checks

lindsayad commented:
If you end up with a mesh with all level 0 elemenets that has hanging nodes, I'd like for there to be an error

GiudGiud committed:
Break out the conformality diagnostics from the diagnostics MG into a util

GiudGiud committed:
Add a check on conformality at the end of the block coarsening operation

GiudGiud committed:
Use right pointer for fixing the mesh

GiudGiud commented:
Added the error. If there's pre-existing non-conformality it will error too, cant discriminate unless I figure out more logic that happens during the coarsening 'sweep'

lindsayad commented on framework/include/utils/MeshBaseDiagnosticsUtils.h:
```diff
@@ -0,0 +1,24 @@
+#pragma once
+
+#include "MooseTypes.h"
+#include "libmesh/mesh_base.h"
```
Can just forward declare the `MeshBase` in namespace `libMesh`

GiudGiud replied:
class MeshBase; ?
or class libMesh::MeshBase?

GiudGiud replied:
both get me an error. first one is 'ambiguous' (two lines later), second says forward declarations cant be nested. 

maybe classMeshBase

then in the declaration
...... libMesh::MeshBase ....)
would work.

lindsayad replied:
```c++
namespace libMesh
{
class MeshBase;
}

void checkNonConformalMesh(const std::unique_ptr<libMesh::MeshBase> & mesh,
```

GiudGiud replied:
done

lindsayad commented on framework/src/utils/MeshBaseDiagnosticsUtils.C:
```diff
@@ -0,0 +1,66 @@
+  auto pl = mesh->sub_point_locator();
+  pl->set_close_to_point_tol(conformality_tol);
+
+  // loop on nodes, assumes a replicated mesh
```
I would error then if `!mesh->is_serial()`

GiudGiud replied:
done

moosebuild commented:
Job [Clang min debug](https://civet.inl.gov/job/2057287/) on 75d05f1 : invalidated by @GiudGiud

unrelated

GiudGiud committed:
Address Alex review
- clean up test requirements

Co-authored-by: Alex Lindsay <alexander.lindsay@inl.gov>

GiudGiud commented:
Thanks for the review!
